design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,679 @@
+  in an attempt to establish parent-child relationships between CRDs.
+
+In addition, the following unimplemented features are goals of the Crossplane
+package manager:
+
+- The ability to resolve dependencies listed in a package for CRDs that do not
+  yet exist in the cluster.
+- The ability for a user to upgrade installed package versions without manual
+  intervention.
+
+As part of these guarantees, Crossplane supports installing packages at both the
+cluster (`ClusterPackageInstall`) and namespace scope (`PackageInstall`). The
+primary difference between these two installation units is that the a
+`ClusterPackageInstall` can only be installed once, and the controller's
+`ServiceAccount` is bound to its `ClusterRole` with a `ClusterRoleBinding`,
+meaning it can watch the resources for which ir requested RBAC at the cluster
+scope. A `PackageInstall`, on the other hand, has its controller's
+`ServiceAccount` bound to its `ClusterRole` with a `RoleBinding` that only
+allows the controller to watch resources in its own namespace.
+
+The advantage of a `PackageInstall` is that is theoretically allows for multiple
+versions of the same package to be installed in a cluster at the same time
+because its controller is only watching for objects in its own namespace.
+However, because CRDs are cluster-scoped objects, there cannot be conflicting
+CRDs installed by two different versions of a `PackageInstall`. So while the
+goal of the `PackageInstall` is to enable multiple versions of a package to
+exist simultaneously in a cluster, the limitations of CRDs makes it less
+effective than desired in practice.
+
+## Goals
+
+The current package infrastructure, though well thought out, has become somewhat
+convoluted and redundant with the introduction of
+[composition](https://github.com/crossplane/crossplane/blob/master/design/design-doc-composition.md)
+into the Crossplane ecosystem.
+
+Composition solves the following goals originally intended to be addressed by a
+`PackageInstall` and [template
+stacks](https://github.com/crossplane/crossplane/blob/master/design/one-pager-template-stacks.md):
+
+- Ability to publish infrastructure abstractions to specific namespaces.
+  - The `PackageInstall` allowed packages to install a namespace-scoped CRD and
+    a controller that only watched for resources in its namespace. Composition
+    enables this by creating a namespace-scoped CRD using an
+    `InfrastructurePublication` and restricting the namespace in which it can be
+    provisioned using RBAC.
+- Ability to create new packages without writing any code.
+  - The templating controller enabled this by allowing a user to create a
+    `behavior.yaml` file and automatically using the templating controller image
+    (configured with the `behavior.yaml`) when one was not supplied in the
+    package. Composition enables this without having to install a new
+    controller. The composition controllers run as part of core Crossplane and
```

Comment:
If composition-based packages depend on a newer version of Crossplane core controllers, will they be able to specify the minimum version of Crossplane their compatible with, e.g. if there's a bugfix they require?

Reply:
This is not currently part of the schema, but I think that is a great candidate to be added!